### PR TITLE
Give a better description of Print using and section variables

### PR DESCRIPTION
**Pull request by jfehrle**
Attn: @Blaisorblade

### CODE REVIEW

File: doc/sphinx/addendum/parallel-proof-processing.rst
Comment (Blaisorblade) [inline doc/sphinx/addendum/parallel-proof-processing.rst:83]: The following would be more accurate (but I'm not too happy with it yet):
```suggestion
Inferred annotations only mention needed variables. Annotations that mention unneeded variables will change the theorem type, and might break existing clients.
```

The reason is that errors might arise never/not at the first reference/have essentially arbitrary error messages.

File: doc/sphinx/proofs/writing-proofs/proof-mode.rst
Comment (Blaisorblade) [inline doc/sphinx/proofs/writing-proofs/proof-mode.rst:376]: `Type` is implied here, so what about:
```suggestion
           Proof using radixNotZero. reflexivity. Qed.
```

**Review by Blaisorblade — CHANGES_REQUESTED** [review]
Thanks for the improvements! I left some suggestions for accuracy.

File: doc/sphinx/addendum/parallel-proof-processing.rst
Comment (jfehrle) [reply doc/sphinx/addendum/parallel-proof-processing.rst:78]: It's best to document things as they are.  We don't know if or when the warning will be fixed.  Also I think mentioning this in the doc may help to motivate a fix--users may ask about it.

File: doc/sphinx/addendum/parallel-proof-processing.rst
Comment (jfehrle) [reply doc/sphinx/addendum/parallel-proof-processing.rst:83]: The text should include the message--that's what the user will see.  What you suggest is much less concrete.  "Inferred annotations" - how are they inferred?  How can you get this not at the first reference?  Can you give an example of breaking an existing client?

Comment (jfehrle) [reply doc/sphinx/addendum/parallel-proof-processing.rst:83]: OK, here's an example:

image

It's rather curious to have a proof take arguments and that the only way to set that up AFAIK is with `Proof using`.  Is that ever useful outside of managing async proofs?  It's unclear how the order of the arguments is determined.

Also, is `All` ever a useful option for `Proof using`?  How about `Type*` or `radixNotZero*`?  Isn't some automatic transitive closure applied even without the `*`?  The doc doesn't specify what relation the transitive closure is applied, either.

Not sure what, if any, of this is worth documenting.

Comment (Blaisorblade) [reply doc/sphinx/addendum/parallel-proof-processing.rst:83]: > The text should include the message--that's what the user will see.

I completely agree, but it should also not say the users things that are not true. What you wrote was generalized from one example. We cannot list all possible errors arising from "your theorem now takes 1 more argument than you expected". Especially, clients might not break when that argument is inferred via type classes.

> It's rather curious to have a proof take arguments and that the only way to set that up AFAIK is with Proof using.

Most theorems take arguments. Sections are just a way to state many theorems over the same arguments, and `Proof using` is only one way to set that up. Try writing statements about `q` and `radix`, and see their inferred types without `Proof.` Here's a made-up example:

```
Lemma square_equal : 0 < q <- q ^ 2 = radix^2 -> q = radix.
Proof. Admitted.
```

After the section, this will take two arguments (`q` and `radix`). BTW, this theorem is false without assuming `radixNotZero` — so `Proof using radixNotZero.` is more appropriate.

> "Inferred annotations" - how are they inferred?

Here's the current description of this inference process. We'll want more editing to make this connection more obvious.

> When a section ends, Coq looks at the proof object to decide which
section variables are actually used and hence have to be quantified in
the statement of the theorem.

Comment (jfehrle) [reply doc/sphinx/addendum/parallel-proof-processing.rst:83]: > I completely agree, but it should also not say the users things that are not true. What you wrote was generalized from one example. We cannot list all possible errors arising from "your theorem now takes 1 more argument than you expected". Especially, clients might not break when that argument is inferred via type classes.

I didn't generalize.  The wording is "an incorrect annotation **_may_** produce a message".  I don't think it's necessary to describe all the possible messages and why they appear.  I can mention other messages if you think they're important enough, but you'd need to provide the basic info on them--me finding them on my own by trial and error takes forever.

I guess the "The error will occur at the first reference to the earlier theorem." should go away?

> Most theorems take arguments.

Implicit, yes, but explicit arguments??

I got this error on your example:

image

You would then apply `square_equal` with a tactic like `apply (square_equal <ARGS>)`?

(Maybe off topic: a long while ago, I was trying to figure out how to use dependent types but even after searching the doc, internet and books over a couple days, I couldn't find the concrete syntax to make it work.  Perhaps this is what I needed?  I was trying to support arrays of length N and assure that subscripts were in bounds.  At some point I'd like to add a doc section on using dependent types in Coq.  But not today.)

> We'll want more editing to make this connection more obvious.

>> When a section ends, Coq looks at the proof object ...

How about adding this example: two variables defined in a section, a lemma references one and not the other, then `Print` the lemma and note that only one is included?

Comment (Blaisorblade) [reply doc/sphinx/addendum/parallel-proof-processing.rst:83]: Just fixed a typo: the thm was `Lemma square_equal : 0 < q -> q ^ 2 = radix^2 -> q = radix.` with `->` not `<-`.

>  I don't think it's necessary to describe all the possible messages and why they appear. I can mention other messages if you think they're important enough, but you'd need to provide the basic info on them--me finding them on my own by trial and error takes forever.

Listing a specific one would suggest that if an error message is completely different, it's unrelated — which is not misleading. I'd rather explain what changes ("a theorem takes unexpected arguments") so people can debug their errors.

Admittedly, I haven't demonstrated the unrelated examples; doing it properly would be some work...

> Implicit, yes, but explicit arguments??

Yep. For instance, `Nat.add_comm` takes two explicit arguments, and `Nat.add_le_mono` takes 6.

```coq
About Nat.add_comm.
About Nat.add_le_mono.
(*
Nat.add_comm : forall n m : nat, n + m = m + n

Nat.add_comm is not universe polymorphic
Arguments Nat.add_comm (n m)%nat_scope
Nat.add_comm is opaque
Expands to: Constant Coq.Arith.PeanoNat.Nat.add_comm
Nat.add_le_mono : forall n m p q : nat, n <= m -> p <= q -> n + p <= m + q

Nat.add_le_mono is not universe polymorphic
Arguments Nat.add_le_mono (n m p q)%nat_scope _ _
Nat.add_le_mono is opaque
Expands to: Constant Coq.Arith.PeanoNat.Nat.add_le_mono
*)
```

It's often hard to tell them apart, since many tactics (like `(e)apply` or `rewrite`) can infer both implicit and explicit arguments — while `Set Implicit Arguments` will silently make some arguments implicit.

> You would then apply square_equal with a tactic like apply (square_equal <ARGS>)?

That's one way; `apply square_equal`, `apply (square_equal <INITIAL_SEQUENCE_OF_ARGS>)`, or just using `square_equal <ARGS>` inside a Gallina term.

**Comment by jfehrle** [comment]
Updated.  I think I addressed your concerns without adding a lot more detail.  My changes are in a new commit, which may make it easier to see what I changed.

File: doc/sphinx/proofs/writing-proofs/proof-mode.rst
Comment (Blaisorblade) [inline doc/sphinx/proofs/writing-proofs/proof-mode.rst:395]: To be sure, `eapply foo’` should also work (and produce a new evar to be solved or goal to be proven).

**Comment by Blaisorblade** [comment]
Cool. I don’t know if I looked closely enough for an approval (and I’m not a doc maintainer anyway, right?), but I think this is already an improvement.

File: doc/sphinx/proofs/writing-proofs/proof-mode.rst
Comment (jfehrle) [reply doc/sphinx/proofs/writing-proofs/proof-mode.rst:395]: Yeah, but let's not get into that here.

**Comment by jfehrle** [comment]
@Blaisorblade Great, thanks.  I expected you would provide the subject matter expertise and that it's up to doc maintainers to do a final pass for typos, grammar, formatting, etc.  Would you approve the PR to remove the "requested changes"?  

@Alizter Would you take a look when convenient and assign?  I will squash just before the merge.

**Comment by Blaisorblade** [comment]
Approving. My only (vague) remaining question is that _some_ of this content is probably also covered in the chapters about sections (https://coq.inria.fr/refman/language/core/sections.html, and the links to https://coq.inria.fr/refman/addendum/type-classes.html#contexts). Maybe we can add links to the https://coq.inria.fr/refman/language/core/sections.html chapter?

**Review by Blaisorblade — APPROVED** [review]

**Comment by jfehrle** [comment]
I added links to the Sections chapter in the Proof Using and Async&Parallel Proofs sections.

In the typeclasses chapter, the only thing that seems relevant is the short section on [Sections and contexts](https://coq.inria.fr/refman/addendum/type-classes.html#sections-and-contexts) that seems like it should be in the Sections chapter.  Beyond that I've never tried to understand the typeclass mechanism, which I suspect would be a bit of a project, not something I want to take on right now.

Updated.

File: doc/sphinx/addendum/parallel-proof-processing.rst
Comment (Alizter) [inline doc/sphinx/addendum/parallel-proof-processing.rst:69]:
```suggestion
      You can either update your proof to not depend on ``precisionNotZero``, or you can
```

File: doc/sphinx/addendum/parallel-proof-processing.rst
Comment (Alizter) [inline doc/sphinx/addendum/parallel-proof-processing.rst:70]:
```suggestion
      update your ``Proof`` line from
```

File: doc/sphinx/addendum/parallel-proof-processing.rst
Comment (Alizter) [inline doc/sphinx/addendum/parallel-proof-processing.rst:71]:
```suggestion
      ``Proof using FtoRradix b pGivesBound precision radix radixMoreThanOne radixMoreThanZERO``
```

File: doc/sphinx/addendum/parallel-proof-processing.rst
Comment (Alizter) [inline doc/sphinx/addendum/parallel-proof-processing.rst:73]:
```suggestion
      ``Proof using FtoRradix b pGivesBound precision precisionNotZero radix radixMoreThanOne radixMoreThanZERO``
```

**Review by Alizter — COMMENTED** [review]
This might not be the correct way to do code blocks, but I noticed some missing code blocks.

File: doc/changelog/07-vernac-commands-and-options/16168-print_using_doc.rst
Comment (jfehrle) [reply doc/changelog/07-vernac-commands-and-options/16168-print_using_doc.rst:2]: Made it ``` :cmd:`Proof using` ```

File: doc/sphinx/addendum/parallel-proof-processing.rst
Comment (jfehrle) [reply doc/sphinx/addendum/parallel-proof-processing.rst:70]: This is literal text for an error message, so I think no change here.

**Comment by jfehrle** [comment]
@Alizter Would you merge this?  @Blaisorblade did a thorough review.  Thanks.

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**
